File: src/core/knowledge/closure.py
"""T-W2-013 知识图谱传递闭包计算模块.

架构 v2 §4.2：闭包预计算后写入 kp_closure 扁平表，热路径查询退化为
单表过滤；递归 CTE 仅在闭包计算时使用（管理查询，非热路径）。

闭包计算规则（任务卡验收 #2）：
- 对 transitive 关系类型（如 prerequisite/composes）：递归展开多跳可达
- 对非 transitive 关系类型（如 confusable）：仅 depth=1 直接边
- path_count：同 (src, dst, rel_type, depth) 的不同路径数

为什么用递归 CTE 而非应用层 BFS：千级节点万级边的图，递归 CTE 在
PostgreSQL 内执行比应用层往返高效；同时利用 DB 端的查询优化器。
循环检测通过 visited 数组实现（防止 acyclic 约束被违反时无限递归）。

宪法 A5/A7：本模块不 import 任何学科包/学段包（学科零特判）。
"""
from __future__ import annotations

from typing import Any

from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from src.core.models.graph_release import GraphRelease
from src.core.models.kp_closure import KpClosure
from src.core.models.relation_type import RelationType


# 闭包深度安全上限——超过此深度判定为环路（违反 acyclic 约束）
_MAX_DEPTH = 50


# 传递闭包递归 CTE：枚举所有路径，再按 (src, dst, depth) 聚合 path_count
# 为什么 visited 数组：避免 acyclic 约束被违反时无限递归（深度爆炸保护）
# 为什么 as_of IS NULL 时不做时间过滤：graph_release.valid_from=NULL 表示
#   "实时快照"（active 状态无时间约束），应包含所有边无论其时间窗如何。
#   仅当 release 携带 valid_from 时才按时间窗过滤（frozen/historical 快照）。
_TRANSITIVE_CLOSURE_SQL = """
WITH RECURSIVE paths AS (
    -- 基础：直接边（depth=1）
    SELECT
        e.src_node_id AS src_node_id,
        e.dst_node_id AS dst_node_id,
        e.rel_type AS rel_type,
        1 AS depth,
        ARRAY[e.src_node_id, e.dst_node_id] AS visited
    FROM kp_edge e
    WHERE e.rel_type = :rel_type
      AND (
        CAST(:as_of AS timestamptz) IS NULL
        OR ((e.valid_from IS NULL OR e.valid_from <= CAST(:as_of AS timestamptz))
            AND (e.valid_to IS NULL OR e.valid_to > CAST(:as_of AS timestamptz)))
      )

    UNION ALL

    -- 递归：在路径末端扩展一跳
    SELECT
        p.src_node_id,
        e.dst_node_id,
        p.rel_type,
        p.depth + 1,
        p.visited || e.dst_node_id
    FROM paths p
    JOIN kp_edge e
      ON e.src_node_id = p.dst_node_id
     AND e.rel_type = p.rel_type
    WHERE p.depth < :max_depth
      AND e.dst_node_id <> ALL(p.visited)  -- 循环检测
      AND (
        CAST(:as_of AS timestamptz) IS NULL
        OR ((e.valid_from IS NULL OR e.valid_from <= CAST(:as_of AS timestamptz))
            AND (e.valid_to IS NULL OR e.valid_to > CAST(:as_of AS timestamptz)))
      )
)
SELECT
    src_node_id,
    dst_node_id,
    rel_type,
    depth,
    COUNT(*) AS path_count
FROM paths
GROUP BY src_node_id, dst_node_id, rel_type, depth
"""


# 非传递关系：仅直接边（depth=1，path_count=1）
_NON_TRANSITIVE_DIRECT_SQL = """
SELECT
    e.src_node_id AS src_node_id,
    e.dst_node_id AS dst_node_id,
    e.rel_type AS rel_type,
    1 AS depth,
    1 AS path_count
FROM kp_edge e
WHERE e.rel_type = :rel_type
  AND (
    CAST(:as_of AS timestamptz) IS NULL
    OR ((e.valid_from IS NULL OR e.valid_from <= CAST(:as_of AS timestamptz))
        AND (e.valid_to IS NULL OR e.valid_to > CAST(:as_of AS timestamptz)))
  )
"""


# 写入闭包条目（按 graph_release_id 关联版本）
_INSERT_CLOSURE_SQL = """
INSERT INTO kp_closure
    (graph_release_id, src_node_id, dst_node_id, rel_type, depth, path_count)
VALUES
    (:graph_release_id, :src_node_id, :dst_node_id, :rel_type, :depth, :path_count)
"""


async def compute_closure(
    graph_release_id: str,
    db: AsyncSession,
    max_depth: int = _MAX_DEPTH,
) -> dict[str, Any]:
    """计算并写入 kp_closure（按 graph_release 版本缓存）.

    幂等：先删除该 graph_release 的既有闭包条目，再重新计算写入。

    Args:
        graph_release_id: 目标图谱版本 id（必须已存在于 graph_release 表）。
        db: AsyncSession（必填）。
        max_depth: 递归深度上限（默认 50，超过判定为环路）。

    Returns:
        统计信息 dict：
        {
            "graph_release_id": str,
            "as_of": datetime | None,
            "closure_rows": int,            # 写入条目总数
            "transitive_rel_types": list[str],
            "non_transitive_rel_types": list[str],
        }

    Raises:
        ValueError: graph_release_id 不存在；max_depth < 1。
    """
    if max_depth < 1:
        raise ValueError(f"max_depth 必须 >= 1，实际 {max_depth}")

    # ── 校验 graph_release 存在并取 valid_from 作为 as-of 时间 ──
    release = await db.get(GraphRelease, graph_release_id)
    if release is None:
        raise ValueError(f"graph_release_id={graph_release_id!r} 不存在")
    as_of = release.valid_from

    # ── 收集所有 relation_type 与其 transitive 标志 ──
    result = await db.execute(text("SELECT rel_type, transitive FROM relation_type"))
    rel_types: list[tuple[str, bool]] = [
        (row[0], bool(row[1])) for row in result.fetchall()
    ]

    # ── 幂等：先删既有闭包条目 ──
    await db.execute(
        text("DELETE FROM kp_closure WHERE graph_release_id = :grid"),
        {"grid": graph_release_id},
    )

    transitive_used: list[str] = []
    non_transitive_used: list[str] = []
    total_rows = 0

    params_common: dict[str, Any] = {
        "graph_release_id": graph_release_id,
        "as_of": as_of,
        "max_depth": max_depth,
    }

    for rel_type, is_transitive in rel_types:
        params = {**params_common, "rel_type": rel_type}
        if is_transitive:
            rows = (await db.execute(text(_TRANSITIVE_CLOSURE_SQL), params)).fetchall()
            transitive_used.append(rel_type)
        else:
            rows = (await db.execute(text(_NON_TRANSITIVE_DIRECT_SQL), params)).fetchall()
            non_transitive_used.append(rel_type)

        # 批量写入
        for row in rows:
            await db.execute(
                text(_INSERT_CLOSURE_SQL),
                {
                    "graph_release_id": graph_release_id,
                    "src_node_id": row[0],
                    "dst_node_id": row[1],
                    "rel_type": row[2],
                    "depth": int(row[3]),
                    "path_count": int(row[4]),
                },
            )
            total_rows += 1

    await db.commit()

    return {
        "graph_release_id": graph_release_id,
        "as_of": as_of,
        "closure_rows": total_rows,
        "transitive_rel_types": transitive_used,
        "non_transitive_rel_types": non_transitive_used,
    }


__all__ = ["compute_closure"]

File: core/knowledge/closure.go
// closure.go 承载知识图谱传递闭包计算（Python 冻结基准
// src/core/knowledge/closure.py 的 Go 移植；T-W2-013）。
//
// 架构 v2 §4.2：闭包预计算后写入 kp_closure 扁平表，热路径查询退化为单表
// 过滤；递归 CTE 仅在闭包计算时使用（管理查询，非热路径）。
//
// 闭包计算规则（任务卡验收 #2）：
//   - 对 transitive 关系类型（如 prerequisite/composes）：递归展开多跳可达；
//   - 对非 transitive 关系类型（如 confusable）：仅 depth=1 直接边；
//   - path_count：同 (src, dst, rel_type, depth) 的不同路径数。
//
// 冻结实现用 PostgreSQL 递归 CTE 枚举所有路径再聚合（千级节点万级边的图在
// DB 内执行优于应用层往返）；Go 移植为纯函数内核（路径枚举 DFS + 计数聚合），
// 语义逐条对齐 CTE：
//   - 基础：直接边（depth=1）；
//   - 递归：在路径末端扩展一跳，深度上限 max_depth（默认 50，超过判定为
//     环路保护——acyclic 约束被违反时的深度爆炸防线）；
//   - 循环检测：CTE 的 visited 数组 → Go 的路径访问集（dst 不得已在路径中）；
//   - 时间窗过滤：as_of 为 nil（release 无 valid_from=实时快照）时不过滤，
//     否则仅保留 valid_from ≤ as_of < valid_to 的边（frozen/historical 快照）。
package knowledge

import (
	"fmt"
	"sort"
	"time"
)

// MaxClosureDepth 闭包深度安全上限——超过此深度判定为环路（违反 acyclic
// 约束；Python _MAX_DEPTH）.
const MaxClosureDepth = 50

// KpEdge 知识点边（kp_edge 表的闭包相关投影；Attrs/Provenance 供种子装载
// 留档）.
type KpEdge struct {
	SrcNodeID  string
	DstNodeID  string
	RelType    string
	ValidFrom  *time.Time
	ValidTo    *time.Time
	Attrs      map[string]any
	Provenance map[string]any
}

// RelationType 关系类型（relation_type 表投影）.
type RelationType struct {
	RelType     string
	Directed    bool
	Transitive  bool
	Acyclic     bool
	Symmetric   bool
	Description *string
}

// GraphRelease 图谱版本（graph_release 表投影；valid_from=NULL 表示实时快照
// （active 状态无时间约束），闭包不做时间过滤）.
type GraphRelease struct {
	ReleaseID    string
	Status       string
	ValidFrom    *time.Time
	ValidTo      *time.Time
	SupersededBy *string
}

// ClosureRow 闭包条目（kp_closure 一行；同 (graph_release_id, src, dst,
// rel_type, depth) 唯一，path_count 承载多路径）.
type ClosureRow struct {
	GraphReleaseID string
	SrcNodeID      string
	DstNodeID      string
	RelType        string
	Depth          int
	PathCount      int
}

// ClosureStats 闭包计算统计（Python compute_closure 返回 dict 的 Go 形）.
type ClosureStats struct {
	GraphReleaseID        string
	AsOf                  *time.Time
	ClosureRows           int
	TransitiveRelTypes    []string
	NonTransitiveRelTypes []string
}

// GraphView 是闭包计算的图快照查询面（对应冻结实现的 graph_release /
// relation_type / kp_edge 三表读取）.
type GraphView interface {
	// Release 取图谱版本；不存在返回错误（Python db.get 返回 None → ValueError）.
	Release(releaseID string) (*GraphRelease, error)
	// RelationTypes 返回全部关系类型及其 transitive 标志.
	RelationTypes() ([]RelationType, error)
	// Edges 返回全部边（含时间窗列）.
	Edges() ([]KpEdge, error)
}

// ClosureStore 是 kp_closure 账面端口（幂等：先删该 release 既有条目再写入，
// Python compute_closure 的 DELETE+INSERT 序）.
type ClosureStore interface {
	// ReplaceClosure 原子替换某 graph_release 的闭包条目；实现方必须强制
	// uq_kpc_release_src_dst_rel_depth / ck_kpc_no_self_loop 等表约束.
	ReplaceClosure(graphReleaseID string, rows []ClosureRow) error
}

// edgeVisible 边的时间窗过滤（CTE 的 valid_from ≤ as_of < valid_to 同义；
// as_of 为 nil 表示实时快照——包含所有边无论其时间窗如何）.
func edgeVisible(e *KpEdge, asOf *time.Time) bool {
	if asOf == nil {
		return true
	}
	if e.ValidFrom != nil && e.ValidFrom.After(*asOf) {
		return false
	}
	if e.ValidTo != nil && !e.ValidTo.After(*asOf) {
		return false
	}
	return true
}

// ComputeClosure 计算并写入 kp_closure（按 graph_release 版本缓存；Python
// compute_closure）。
//
// 幂等：先删除该 graph_release 的既有闭包条目，再重新计算写入。
// maxDepth < 1 或 release 不存在时返回错误。
func ComputeClosure(graphReleaseID string, graph GraphView, store ClosureStore, maxDepth int) (*ClosureStats, error) {
	if maxDepth < 1 {
		return nil, fmt.Errorf("knowledge: max_depth 必须 >= 1，实际 %d", maxDepth)
	}
	// ── 校验 graph_release 存在并取 valid_from 作为 as-of 时间 ──
	release, err := graph.Release(graphReleaseID)
	if err != nil {
		return nil, err
	}
	if release == nil {
		return nil, fmt.Errorf("knowledge: graph_release_id=%q 不存在", graphReleaseID)
	}
	asOf := release.ValidFrom

	// ── 收集所有 relation_type 与其 transitive 标志 ──
	relTypes, err := graph.RelationTypes()
	if err != nil {
		return nil, err
	}
	// 遍历序确定性：按 rel_type 排序（冻结实现为 DB 行序；键序同样可复现）
	ordered := append([]RelationType(nil), relTypes...)
	sort.Slice(ordered, func(i, j int) bool { return ordered[i].RelType < ordered[j].RelType })

	edges, err := graph.Edges()
	if err != nil {
		return nil, err
	}

	transitiveUsed := []string{}
	nonTransitiveUsed := []string{}
	allRows := []ClosureRow{}

	for _, rt := range ordered {
		var rows []ClosureRow
		if rt.Transitive {
			rows, err = transitiveClosureRows(graphReleaseID, rt.RelType, edges, asOf, maxDepth)
			transitiveUsed = append(transitiveUsed, rt.RelType)
		} else {
			rows, err = nonTransitiveDirectRows(graphReleaseID, rt.RelType, edges, asOf)
			nonTransitiveUsed = append(nonTransitiveUsed, rt.RelType)
		}
		if err != nil {
			return nil, err
		}
		allRows = append(allRows, rows...)
	}

	// ── 幂等写入（先删后插由 store 承担）──
	if err := store.ReplaceClosure(graphReleaseID, allRows); err != nil {
		return nil, err
	}

	return &ClosureStats{
		GraphReleaseID:        graphReleaseID,
		AsOf:                  asOf,
		ClosureRows:           len(allRows),
		TransitiveRelTypes:    transitiveUsed,
		NonTransitiveRelTypes: nonTransitiveUsed,
	}, nil
}

// transitiveClosureRows 传递闭包：路径枚举 + 聚合（_TRANSITIVE_CLOSURE_SQL
// 的递归 CTE 同义实现）。
//
// CTE 形态：基础 = depth 1 的直接边（visited=[src,dst]）；递归 = 在路径末端
// 扩展一跳（depth+1，dst 不得已在 visited 中，depth < max_depth 才续）；
// 最终按 (src, dst, rel_type, depth) GROUP BY COUNT(*) 为 path_count。
func transitiveClosureRows(graphReleaseID, relType string, edges []KpEdge, asOf *time.Time, maxDepth int) ([]ClosureRow, error) {
	// 邻接表：仅该 rel_type 且时间窗可见的边；目标序排序保证枚举确定性.
	adj := map[string][]string{}
	for i := range edges {
		e := &edges[i]
		if e.RelType != relType || !edgeVisible(e, asOf) {
			continue
		}
		adj[e.SrcNodeID] = append(adj[e.SrcNodeID], e.DstNodeID)
	}
	for src := range adj {
		targets := adj[src]
		sort.Strings(targets)
		adj[src] = targets
	}

	counts := map[closureKey]int{}
	// 以每条可见直接边为路径种子，DFS 枚举全部简单路径.
	var walk func(src, current string, depth int, visited map[string]struct{})
	walk = func(src, current string, depth int, visited map[string]struct{}) {
		counts[closureKey{src, current, depth}]++
		if depth >= maxDepth {
			return // p.depth < :max_depth 才允许继续扩展
		}
		for _, next := range adj[current] {
			if _, inPath := visited[next]; inPath {
				continue // e.dst_node_id <> ALL(p.visited)：循环检测
			}
			visited[next] = struct{}{}
			walk(src, next, depth+1, visited)
			delete(visited, next)
		}
	}
	for src := range adj {
		for _, dst := range adj[src] {
			visited := map[string]struct{}{src: {}, dst: {}}
			walk(src, dst, 1, visited)
		}
	}

	rows := make([]ClosureRow, 0, len(counts))
	for k, n := range counts {
		if k.src == k.dst {
			// ck_kpc_no_self_loop：src<>dst（CTE 对自环边会产出 A→A 行，
			// 冻结实现在 INSERT 时触发 CHECK 失败——这里前置为显式错误）
			return nil, fmt.Errorf("knowledge: 闭包出现自环 src=dst=%q（rel_type=%s；违反 ck_kpc_no_self_loop，请检查图数据 acyclic 约束）", k.src, relType)
		}
		rows = append(rows, ClosureRow{
			GraphReleaseID: graphReleaseID,
			SrcNodeID:      k.src,
			DstNodeID:      k.dst,
			RelType:        relType,
			Depth:          k.depth,
			PathCount:      n,
		})
	}
	sortClosureRows(rows)
	return rows, nil
}

// nonTransitiveDirectRows 非传递关系：仅直接边（depth=1；每条可见边一条，
// path_count=1——_NON_TRANSITIVE_DIRECT_SQL 同义）.
func nonTransitiveDirectRows(graphReleaseID, relType string, edges []KpEdge, asOf *time.Time) ([]ClosureRow, error) {
	type key struct{ src, dst string }
	counts := map[key]int{}
	for i := range edges {
		e := &edges[i]
		if e.RelType != relType || !edgeVisible(e, asOf) {
			continue
		}
		counts[key{e.SrcNodeID, e.DstNodeID}]++
	}
	rows := make([]ClosureRow, 0, len(counts))
	for k, n := range counts {
		if k.src == k.dst {
			return nil, fmt.Errorf("knowledge: 闭包出现自环 src=dst=%q（rel_type=%s；违反 ck_kpc_no_self_loop，请检查图数据 acyclic 约束）", k.src, relType)
		}
		// uq_kpc_release_src_dst_rel_depth：同 (src,dst) 并行边在 kp_edge 的
		// uq_kp_edge_src_dst_rel 下不可能；出现即数据错，显式报错而非静默聚合.
		if n > 1 {
			return nil, fmt.Errorf("knowledge: rel_type=%s 存在并行边 %s→%s（%d 条）；违反 uq_kp_edge_src_dst_rel", relType, k.src, k.dst, n)
		}
		rows = append(rows, ClosureRow{
			GraphReleaseID: graphReleaseID,
			SrcNodeID:      k.src,
			DstNodeID:      k.dst,
			RelType:        relType,
			Depth:          1,
			PathCount:      1,
		})
	}
	sortClosureRows(rows)
	return rows, nil
}

// closureKey 聚合键 (src, dst, depth)（rel_type 在调用方作用域内恒定）.
type closureKey struct {
	src   string
	dst   string
	depth int
}

// sortClosureRows 确定性输出序：(rel_type, src, dst, depth).
func sortClosureRows(rows []ClosureRow) {
	sort.Slice(rows, func(i, j int) bool {
		if rows[i].RelType != rows[j].RelType {
			return rows[i].RelType < rows[j].RelType
		}
		if rows[i].SrcNodeID != rows[j].SrcNodeID {
			return rows[i].SrcNodeID < rows[j].SrcNodeID
		}
		if rows[i].DstNodeID != rows[j].DstNodeID {
			return rows[i].DstNodeID < rows[j].DstNodeID
		}
		return rows[i].Depth < rows[j].Depth
	})
}
